File: pgpainless-core/src/main/java/org/pgpainless/key/SubkeyIdentifier.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key;

import java.util.NoSuchElementException;
import javax.annotation.Nonnull;

import org.bouncycastle.openpgp.PGPKeyRing;
import org.bouncycastle.openpgp.PGPPublicKey;

/**
 * Tuple class used to identify a subkey by fingerprints of the primary key of the subkeys key ring,
 * as well as the subkeys fingerprint.
 */
public class SubkeyIdentifier {

    private final OpenPgpV4Fingerprint primaryKeyFingerprint;
    private final OpenPgpV4Fingerprint subkeyFingerprint;

    /**
     * Create a {@link SubkeyIdentifier} from a {@link PGPKeyRing}.
     * The identifier will point to the primary key of the provided ring.
     *
     * @param keyRing key ring
     */
    public SubkeyIdentifier(PGPKeyRing keyRing) {
        this(keyRing, keyRing.getPublicKey().getKeyID());
    }

    /**
     * Create a {@link SubkeyIdentifier} from a {@link PGPKeyRing} and the subkeys key id.
     * {@link #getPrimaryKeyFingerprint()} will return the {@link OpenPgpV4Fingerprint} of the keyrings primary key,
     * while {@link #getSubkeyFingerprint()} will return the subkeys fingerprint.
     *
     * @param keyRing keyring the subkey belongs to
     * @param keyId keyid of the subkey
     */
    public SubkeyIdentifier(@Nonnull PGPKeyRing keyRing, long keyId) {
        PGPPublicKey subkey = keyRing.getPublicKey(keyId);
        if (subkey == null) {
            throw new NoSuchElementException("Key ring does not contain subkey with id " + Long.toHexString(keyId));
        }
        this.primaryKeyFingerprint = new OpenPgpV4Fingerprint(keyRing);
        this.subkeyFingerprint = new OpenPgpV4Fingerprint(subkey);
    }

    public SubkeyIdentifier(@Nonnull PGPKeyRing keyRing, @Nonnull OpenPgpV4Fingerprint subkeyFingerprint) {
        this(new OpenPgpV4Fingerprint(keyRing), subkeyFingerprint);
    }

    /**
     * Create a {@link SubkeyIdentifier} that identifies the primary key with the given fingerprint.
     * This means, both {@link #getPrimaryKeyFingerprint()} and {@link #getSubkeyFingerprint()} return the same.
     *
     * @param primaryKeyFingerprint fingerprint of the identified key
     */
    public SubkeyIdentifier(@Nonnull OpenPgpV4Fingerprint primaryKeyFingerprint) {
        this(primaryKeyFingerprint, primaryKeyFingerprint);
    }

    /**
     * Create a {@link SubkeyIdentifier} which points to the subkey with the given subkeyFingerprint,
     * which has a primary key with the given primaryKeyFingerprint.
     *
     * @param primaryKeyFingerprint fingerprint of the primary key
     * @param subkeyFingerprint fingerprint of the subkey
     */
    public SubkeyIdentifier(@Nonnull OpenPgpV4Fingerprint primaryKeyFingerprint, @Nonnull OpenPgpV4Fingerprint subkeyFingerprint) {
        this.primaryKeyFingerprint = primaryKeyFingerprint;
        this.subkeyFingerprint = subkeyFingerprint;
    }

    public @Nonnull OpenPgpV4Fingerprint getFingerprint() {
        return getSubkeyFingerprint();
    }

    public long getKeyId() {
        return getSubkeyId();
    }

    /**
     * Return the {@link OpenPgpV4Fingerprint} of the primary key of the identified key.
     * This might be the same as {@link #getSubkeyFingerprint()} if the identified subkey is the primary key.
     *
     * @return primary key fingerprint
     */
    public @Nonnull OpenPgpV4Fingerprint getPrimaryKeyFingerprint() {
        return primaryKeyFingerprint;
    }

    /**
     * Return the key id of the primary key of the identified key.
     * This might be the same as {@link #getSubkeyId()} if the identified subkey is the primary key.
     *
     * @return primary key id
     */
    public long getPrimaryKeyId() {
        return getPrimaryKeyFingerprint().getKeyId();
    }

    /**
     * Return the {@link OpenPgpV4Fingerprint} of the identified subkey.
     *
     * @return subkey fingerprint
     */
    public @Nonnull OpenPgpV4Fingerprint getSubkeyFingerprint() {
        return subkeyFingerprint;
    }

    /**
     * Return the key id of the identified subkey.
     *
     * @return subkey id
     */
    public long getSubkeyId() {
        return getSubkeyFingerprint().getKeyId();
    }

    @Override
    public int hashCode() {
        return primaryKeyFingerprint.hashCode() * 31 + subkeyFingerprint.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof SubkeyIdentifier)) {
            return false;
        }
        SubkeyIdentifier other = (SubkeyIdentifier) obj;
        return getPrimaryKeyFingerprint().equals(other.getPrimaryKeyFingerprint())
                && getSubkeyFingerprint().equals(other.getSubkeyFingerprint());
    }

    @Override
    public String toString() {
        return getSubkeyFingerprint() + " " + getPrimaryKeyFingerprint();
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/key/SubkeyIdentifier.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key

import org.bouncycastle.bcpg.KeyIdentifier
import org.bouncycastle.openpgp.PGPKeyRing
import org.bouncycastle.openpgp.PGPPublicKey
import org.bouncycastle.openpgp.api.OpenPGPCertificate.OpenPGPComponentKey
import org.bouncycastle.openpgp.api.OpenPGPKey.OpenPGPPrivateKey

/**
 * Tuple class used to identify a subkey (component key) by fingerprints of the certificate, as well
 * as the component keys fingerprint.
 */
class SubkeyIdentifier(
    /**
     * Fingerprint of the certificate.
     */
    val certificateFingerprint: OpenPgpFingerprint,
    /**
     * Fingerprint of the target component key.
     */
    val componentKeyFingerprint: OpenPgpFingerprint
) {

    /**
     * Constructor for a [SubkeyIdentifier] pointing to the primary key identified by the
     * [certificateFingerprint].
     *
     * @param certificateFingerprint primary key fingerprint
     */
    constructor(
        certificateFingerprint: OpenPgpFingerprint
    ) : this(certificateFingerprint, certificateFingerprint)

    /**
     * Constructor for a [SubkeyIdentifier] pointing to the primary key of the given [PGPKeyRing].
     *
     * @param certificate certificate
     */
    constructor(certificate: PGPKeyRing) : this(certificate.publicKey)

    /**
     * Constructor for a [SubkeyIdentifier] pointing to the given [primaryKey].
     *
     * @param primaryKey primary key
     */
    constructor(primaryKey: PGPPublicKey) : this(OpenPgpFingerprint.of(primaryKey))

    /**
     * Constructor for a [SubkeyIdentifier] pointing to a component key (identified by
     * [componentKeyId]) from the given [certificate].
     */
    @Deprecated("Pass in a KeyIdentifier instead of a keyId.")
    constructor(
        certificate: PGPKeyRing,
        componentKeyId: Long
    ) : this(certificate, KeyIdentifier(componentKeyId))

    /**
     * Constructor for a [SubkeyIdentifier] pointing to the given [componentKey].
     *
     * @param componentKey component key
     */
    constructor(
        componentKey: OpenPGPComponentKey
    ) : this(OpenPgpFingerprint.of(componentKey.certificate), OpenPgpFingerprint.of(componentKey))

    /** Constructor for a [SubkeyIdentifier] pointing to the given [componentKey]. */
    constructor(componentKey: OpenPGPPrivateKey) : this(componentKey.secretKey)

    /**
     * Constructor for a [SubkeyIdentifier] pointing to a component key (identified by the
     * [componentKeyFingerprint]) of the given [certificate].
     *
     * @param certificate certificate
     * @param componentKeyFingerprint fingerprint of the component key
     */
    constructor(
        certificate: PGPKeyRing,
        componentKeyFingerprint: OpenPgpFingerprint
    ) : this(OpenPgpFingerprint.of(certificate), componentKeyFingerprint)

    /**
     * Constructor for a [SubkeyIdentifier] pointing to a component key (identified by the
     * [componentKeyIdentifier]) of the given [certificate].
     *
     * @param certificate certificate
     * @param componentKeyIdentifier identifier of the component key
     */
    constructor(
        certificate: PGPKeyRing,
        componentKeyIdentifier: KeyIdentifier
    ) : this(
        OpenPgpFingerprint.of(certificate),
        OpenPgpFingerprint.of(
            certificate.getPublicKey(componentKeyIdentifier)
                ?: throw NoSuchElementException(
                    "OpenPGP key does not contain subkey $componentKeyIdentifier")))

    @Deprecated(
        "Use certificateFingerprint instead.", replaceWith = ReplaceWith("certificateFingerprint"))
    val primaryKeyFingerprint: OpenPgpFingerprint = certificateFingerprint

    @Deprecated(
        "Use componentKeyFingerprint instead.",
        replaceWith = ReplaceWith("componentKeyFingerprint"))
    val subkeyFingerprint: OpenPgpFingerprint = componentKeyFingerprint

    /** [KeyIdentifier] of the component key. */
    val keyIdentifier = componentKeyFingerprint.keyIdentifier

    /** [KeyIdentifier] of the component key. */
    val componentKeyIdentifier = keyIdentifier

    /** [KeyIdentifier] of the primary key of the certificate the component key belongs to. */
    val certificateIdentifier = certificateFingerprint.keyIdentifier

    /** Key-ID of the component key. */
    @Deprecated("Use of key-ids is discouraged.") val keyId = keyIdentifier.keyId

    /** Fingerprint of the component key. */
    val fingerprint = componentKeyFingerprint

    /** Key-ID of the component key. */
    @Deprecated("Use of key-ids is discouraged.") val subkeyId = componentKeyIdentifier.keyId

    /** Key-ID of the primary key of the certificate the component key belongs to. */
    @Deprecated("Use of key-ids is discouraged.") val primaryKeyId = certificateIdentifier.keyId

    /** True, if the component key is the primary key. */
    val isPrimaryKey = certificateIdentifier.matches(componentKeyIdentifier)

    /**
     * Return true, if the provided [fingerprint] matches either the [certificateFingerprint]
     * or [componentKeyFingerprint].
     */
    fun matches(fingerprint: OpenPgpFingerprint) =
        certificateFingerprint == fingerprint || componentKeyFingerprint == fingerprint

    override fun equals(other: Any?): Boolean {
        if (other == null) {
            return false
        }
        if (this === other) {
            return true
        }
        if (other !is SubkeyIdentifier) {
            return false
        }

        return certificateFingerprint == other.certificateFingerprint &&
            componentKeyFingerprint == other.componentKeyFingerprint
    }

    override fun hashCode(): Int {
        return certificateFingerprint.hashCode() + 31 * componentKeyFingerprint.hashCode()
    }

    override fun toString(): String = "$componentKeyFingerprint $certificateFingerprint"
}
